src/main/java/com/redhat/jenkins/plugins/cachet/GlobalCachetConfiguration.java

```diff
@@ -39,38 +42,49 @@
 @Extension
 public final class GlobalCachetConfiguration extends GlobalConfiguration {
 
-    private String cachetUrl;
-    private Boolean ignoreSSL;
-
-    public boolean getIgnoreSSL() {
-        return BooleanUtils.toBoolean(ignoreSSL);
-    }
-
-    @DataBoundSetter
-    public void setIgnoreSSL(Boolean ignoreSSL) {
-        this.ignoreSSL = ignoreSSL;
-    }
+    private List<SourceTemplate> sources = new ArrayList<>();
 
+    @DataBoundConstructor
     public GlobalCachetConfiguration() {
         load();
     }
 
-    public String getCachetUrl() {
-        return cachetUrl;
+    @DataBoundSetter
+    public void setSources(List<SourceTemplate> sources) {
+        this.sources = sources;
     }
 
-    @DataBoundSetter
-    public void setCachetUrl(String cachetUrl) {
-        this.cachetUrl = cachetUrl;
+    public List<SourceTemplate> getSources() {
+        return sources != null ? sources : Collections.emptyList();
     }
 
+    @Nonnull
     @Override
     public String getDisplayName() {
         return Messages.pluginName();
     }
 
     @Override
     public boolean configure(StaplerRequest req, JSONObject json) {
+        HashMap<String, String> urls = new HashMap<>();
```

Comment:
Do not use configure to accept form submissions, rely on `@DataBoundConstructor`s (you are declaring anyway) instead.

Reply:
I'm using configure to check for unique urls & labels, is there another recommended way of doing it?

Reply:
Yeah, the unfortunate part is the JSON is first parsed by you manually to do the check and then by stapler's `bindJson`. It would be cleaner to move the uniqueness-enforcing logic into `@DataBoundSetter`/`@DataBoundConstrucotr` of `GlobalCachetConfiguration` so you can focus on enforcing invariants ale benefit from the fact stapler has already build your objects from JSON. The exception thrown from there should have the same effect.

Reply:
Discussing this in detail, it turns out this is not as easy as I hoped. `@DataBoundConstructor` is not called on submission as the class is a singleton, and if the validation gets moved to the `setSources` it causes problems as the method does not get called when the list if changed to an empty one. I am ok to keep this the way it is.